lua/pac3/core/client/parts/interpolated_multibone.lua

```diff
@@ -0,0 +1,230 @@
+--STAGE			0		1		2		3
+--PROPORTION	0	0.5	0	0.5	0	0.5	3
+function PART:OnDraw()
+	self:UpdateNodes()
+	if self.valid_time > CurTime() then return end
+	
+	self.pos = self.pos or self:GetWorldPosition()
+	self.ang = self.ang or self:GetWorldAngles()
+	
+	if not self.Preview then hook.Remove("PostDrawOpaqueRenderables", "Multibone_draw"..self.UniqueID) end
+
+	local stage = math.max(0,math.floor(self.LerpValue))
+	local proportion = math.max(0,self.LerpValue) % 1
+
+	if self.Preview then
+		hook.Add("PostDrawOpaqueRenderables", "Multibone_draw"..self.UniqueID, function()
+			render.DrawLine(self.pos,self.pos + self.ang:Forward()*50, Color(255,0,0))
+			render.DrawLine(self.pos,self.pos - self.ang:Right()*50, Color(0,255,0))
+			render.DrawLine(self.pos,self.pos + self.ang:Up()*50, Color(0,0,255))
+			render.DrawWireframeSphere(self.pos, 8 + 2*math.sin(5*RealTime()), 15, 15, Color(255,255,255), true)
+		end)
+	end
+	self:Interpolate(stage,proportion)
+	
+end
+
+function PART:UpdateNodes()
+	for i=1,10,1 do
+		self.nodes["Node"..i] = self["Node"..i]
+	end
+end
+
+function PART:SetWorldPos(x,y,z)
+	self.pos.x = x
+	self.pos.y = y
+	self.pos.z = z
+end
+
+function PART:Interpolate(stage, proportion)
+	--print("Calculated the stage. We are at stage " .. stage .. " between nodes " .. stage .. " and " .. (stage + 1))
+	local firstnode
+	if stage <= 0 then
+		firstnode = self
+	else
+		firstnode = self.nodes["Node"..stage] or self
+	end
+	
+	
+	local secondnode = self.nodes["Node"..stage+1]
+	if firstnode == nil or firstnode == NULL or not firstnode.GetWorldPosition then firstnode = self end
+	if secondnode == nil or secondnode == NULL or not secondnode.GetWorldPosition then secondnode = self end
+
+	proportion = math.pow(proportion,self.Power)
+	if secondnode ~= nil and secondnode ~= NULL then
+		self.pos = (1-proportion)*(firstnode:GetWorldPosition()) + (secondnode:GetWorldPosition())*proportion
+		self.ang = GetClosestAngleMidpoint(firstnode:GetWorldAngles(), secondnode:GetWorldAngles(), proportion)
+		--self.ang = (1-proportion)*(firstnode:GetWorldAngles() + Angle(360,360,360)) + (secondnode:GetWorldAngles() + Angle(360,360,360))*proportion
+	elseif proportion == 0 then
+		self.pos = firstnode:GetWorldPosition()
+		self.ang = firstnode:GetWorldAngles()
+	else
+		if self.InterpolatePosition then self.pos = (1-proportion)*self:GetWorldPosition() + (self:GetWorldPosition())*proportion end
+		if self.InterpolateAngles then self.ang = GetClosestAngleMidpoint(self:GetWorldAngles(), self:GetWorldAngles(), proportion) end
+		--self.ang = (1-proportion)*(self:GetWorldAngles() + Angle(360,360,360)) + (self:GetWorldAngles() + Angle(360,360,360))*proportion
+	end
+
+	if not self.InterpolatePosition then
+		self.pos = self:GetWorldPosition()
+	end
+	if not self.InterpolateAngles then
+		self.ang = self:GetWorldAngles()
+	end
+	self.Owner:SetPos(self.pos)
+	self.Owner:SetAngles(self.ang)
+end
+
+function GetClosestAngleMidpoint(a1, a2, proportion)
```

Comment:
Angle interpolation should be a rotation axis between the two angle endpoints of length (rotation radians * proportion). We shouldn't be interpolating yaw pitch roll values. I don't remember if glua has a slerp function that can do this but sad if not. If you need code for this lemme know.

Reply:
I think I just got ahead of myself out of wanting to implement it myself, it didn't end up working perfectly because of roll and some oddities, but it was something. I will use LerpVector and LerpAngle.